fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/ModDiscoverer.java

```diff
@@ -82,7 +84,17 @@ public ModValidator discoverMods() {
                 var exceptions = candidates.stream().map(IModLocator.ModFileOrException::ex).filter(Objects::nonNull).toList();
                 if (!exceptions.isEmpty()) {
                     LOGGER.debug(LogMarkers.SCAN, "Locator {} found {} invalid mod files", locator, exceptions.size());
-                    brokenFiles.addAll(exceptions.stream().map(e->e instanceof InvalidModFileException ime ? ime.getBrokenFile() : null).filter(Objects::nonNull).toList());
+                    for (ModFileLoadingException ex : exceptions) {
+                        // pipe exception messages through the discoveryErrorData to avoid swallowing some exceptions and improve error messages
+                        // (no longer a generic "Invalid mod file" for all InvalidModExceptions - it actually shows the exception message now)
+                        if (ex instanceof InvalidModFileException invalidModFileEx) {
+                            var modInfo = invalidModFileEx.getBrokenFile();
+                            brokenFiles.add(modInfo);
+                            discoveryErrorData.add(new EarlyLoadingException.ExceptionData(ex.getMessage(), modInfo));
+                        } else {
+                            discoveryErrorData.add(new EarlyLoadingException.ExceptionData(ex.getMessage()));
+                        }
+                    }
```

Comment:
ok, so this new data we're creating here in `discoveryErrorData` gets sent to the ModInfo instance. sucks that the exception has data specific to it because that errordata.add line looks so similar in both branches.

Reply:
Passing in a null `modInfo` actually worked, but I didn't want to risk a possible breaking change (I have no idea if the context objects are expected to be non-null as the codebase is a mix of nullable and non-nullable - often with no clear indication of which)

Reply:
if that's an issue with forge's code base, you should add clarification of nullability annotations in the forge 2024 wishlist